src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,316 @@
+
+    /* The set B from Johnson's paper */
+    igraph_adjlist_t B;
+
+    /* Stack in which the vertices of the current cycle are pushed */
+    igraph_vector_int_t stack;
+
+    /* Boolean vector indicating which vertices are blocked */
+    igraph_vector_bool_t blocked;
+
+    /* Whether the graph is directed */
+    igraph_bool_t directed;
+} igraph_simple_cycle_search_state_t;
+
+/* The implementation of procedure UNBLOCK from Johnson's paper */
+
+static igraph_error_t igraph_i_simple_cycles_unblock(
+    igraph_simple_cycle_search_state_t *state, igraph_integer_t u
+) {
+    igraph_vector_int_t* neis;
+    igraph_integer_t w;
+
+    VECTOR(state->blocked)[u] = false;
+
+    neis = igraph_adjlist_get(&state->B, u);
+    while (!igraph_vector_int_empty(neis)) {
+        w = igraph_vector_int_pop_back(neis);
+        if (VECTOR(state->blocked)[w]) {
+            IGRAPH_CHECK(igraph_i_simple_cycles_unblock(state, w));
+        }
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+/* The implementation of procedure CIRCUIT from Johnson's paper */
+
+static igraph_error_t igraph_i_simple_cycles_circuit(
+    igraph_simple_cycle_search_state_t *state, igraph_integer_t V,
+    igraph_integer_t S, igraph_vector_int_list_t *results, bool *found,
+    igraph_simple_cycle_search_mode_t search_mode
+) {
+    bool local_found = false;
+    igraph_vector_int_t* neighbors;
+    igraph_integer_t num_neighbors;
+
+    // stack v
+    IGRAPH_CHECK(igraph_vector_int_push_back(&state->stack, V));
+    // printf("Pushing %lld to stack, stack size is %lld, result size is %lld\n", V, igraph_vector_int_size(&state->stack), igraph_vector_int_list_size(results));
+    VECTOR(state->blocked)[V] = true;
+
+    // L1
+    neighbors = igraph_adjlist_get(&state->AK, V);
+    num_neighbors = igraph_vector_int_size(neighbors);
+    for (igraph_integer_t i = 0; i < num_neighbors; ++i) {
+        igraph_integer_t W = VECTOR(*neighbors)[i];
+        // NOTE: possibly dangerous fix for undirected graphs,
+        // disabling finding any two-vertex-loops
+        if (W == S) {
+            if ((state->directed || igraph_vector_int_size(&state->stack) > 2)) {
+                local_found = true;
+                // output circuit composed of stack
+                // printf("Found cycle with size %" IGRAPH_PRId "\n", igraph_vector_int_size(&state->stack));
+
+                // copy output: from stack to vector
+                igraph_vector_int_t res;
+                IGRAPH_CHECK(igraph_vector_int_init_copy(&res, &state->stack));
+                IGRAPH_FINALLY(igraph_vector_int_destroy, &res);
+                IGRAPH_CHECK(igraph_vector_int_reverse(&res));
+                // undirected graphs lead to every cycle being found twice.
+                // this is our naïve filter for now
+                igraph_bool_t persist_result = true;
+                if (!state->directed && search_mode == IGRAPH_UNDIRECTED_CYCLE_SEARCH_ONE) {
+                    igraph_vector_int_sort(&res);
```

Comment:
This line here kinda bothers me because it means that for undirected graphs with `search_mode == IGRAPH_UNDIRECTED_CYCLE_SEARCH_ONE`, the algorithm will return the vertices in the cycle ordered by their vertex IDs, not in the order they are to be found in the cycle -- but for all other cases (other search modes, directed graphs) it will return the vertices in order.

I'll probably refactor this part of the code to something that does not involve sorting the vector.

Reply:
I agree, it is a bad solution. Ideally, there would be an easy way to compute a hash for each result, and just check whether that hash was found already. 

The straight-forward way to circumvent the problem you mention would be to copy the vector and keep both the sorted and unsorted versions, but given that this algorithm might lead to memory issues for large/lattice like graphs anyway, I would be somewhat cautios with that.

Reply:
Just noting that we have a function that orders edges (not vertices!) in cycle order https://github.com/igraph/igraph/blob/master/src/misc/order_cycle.cpp#L34